### PR TITLE
Rq dashboard extension

### PR DESCRIPTION
**Pull request by ibrahim12**
### Feature Improvements
- Show state tables for queue (running, pending, finished) along side job stop/cancel/requeue/delete option.
- Option to show jobs in reverse order
- Hide workers by default
- Hide Failed jobs traceback by default and toggle on click
### Code Improvements
- Refactored jinja2 templates for re-using views.

### CODE REVIEW

File: dev_scripts/test.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+from redis import Redis
```
Comment (lucasnad27) [inline dev_scripts/test.py]: this should be part of the test suite, not hidden away in `dev_scripts`

File: dev_scripts/test.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+
+def create_n_jobs_on_queue(n, queue, fn):
+    for index in range(n):
+        queue.enqueue(fn, index, ttl=3600, result_ttl=3600)
+
+
+# create_n_jobs_on_queue(10, default, functions.func_default)
+create_n_jobs_on_queue(10, new, functions.func_default)
+# create_n_jobs_on_queue(10, new, functions.func_with_exception)
```
Comment (lucasnad27) [inline dev_scripts/test.py]: No commented out code please

File: rq_dashboard/default_settings.py [hunk 3]
```diff
@@ -19,3 +19,5 @@
 DEBUG = False
 RQ_DASHBOARD_WEB_BACKGROUND = "black"
 RQ_DASHBOARD_DELETE_JOBS = False
+
+RQ_DASHBOARD_JOB_SORT_ORDER = '-age'
```
Comment (lucasnad27) [inline rq_dashboard/default_settings.py]: Nice setting here!

File: rq_dashboard/templates/rq_dashboard/dashboard.js [hunk 4]
```diff
@@ -6,11 +6,15 @@ var url_for = function(name, param) {
     else if (name == 'workers') { url += 'workers.json'; }
     else if (name == 'cancel_job') { url += 'job/' + encodeURIComponent(param) + '/cancel'; }
     else if (name == 'requeue_job') { url += 'job/' + encodeURIComponent(param) + '/requeue'; }
+    else if (name == 'requeue_finished_job') { url += 'job/' + encodeURIComponent(param) +  '/finished/requeue'; }
```
Comment (lucasnad27) [inline rq_dashboard/templates/rq_dashboard/dashboard.js]: Why would we want to requeue a finished job?

File: rq_dashboard/templates/rq_dashboard/dashboard.js [hunk 5]
```diff
@@ -226,212 +232,14 @@ var modalConfirm = function(action, cb) {
...
-            $.post($this.attr('href'), function(data) {});
-        });
-        return false;
-    });
-
-    // Enable the AJAX behaviour of the cancel button
-    $tbody.on('click', '[data-role=cancel-job-btn]', function(e) {
-        e.preventDefault();
-        e.stopPropagation();
+{{ job_table.render_js(queue.name, 'pending', page if (state == 'pending' or queue_name == 'failed') else 1)  }}
```
Comment (lucasnad27) [inline rq_dashboard/templates/rq_dashboard/dashboard.js]: Stopping my review here as I need to attend to other items. but will continue reviewing later. I want to give you my feedback up to this point.

**Review by lucasnad27 — CHANGES_REQUESTED** [review]
In general, I don't like having a bunch of "random" developer code living in the `dev_scripts` directory. It's not a well understood practice/pattern. If you need to add helpers for local dev (e.g., boot worker) it should go into the Makefile. You could also make the case for putting it into the fabfile.py

File: rq_dashboard/templates/rq_dashboard/dashboard.js (hunk 4 above)
Comment (ibrahim12) [reply rq_dashboard/templates/rq_dashboard/dashboard.js]: Running that job again ?

File: dev_scripts/test.py (hunk 1 above)
Comment (ibrahim12) [reply dev_scripts/test.py]: I will make a fab script

Comment (ibrahim12) [reply dev_scripts/test.py]: removed

**Review by lucasnad27 — CHANGES_REQUESTED** [review]
I noticed two commits from the upstream maintainer. We should _NOT_ spend our time reviewing the maintainer's code, unless we're looking at updating our fork. But that should be done outside of the context of our own changes and pull requests. Here is the recommended flow:
1) git checkout master
2) git fetch --upstream
3) git merge upstream/master

If you have a feature branch that needs the most recent changes from the upstream branch, you should rebase*  from origin/master (our version of master) after you've completed the steps from above.

* or merge if that's a foreign concept to you. Hint, hint...rebasing is a great skill in your tool belt to have!

Sorry I'm being a stickler on this, but if we're going to maintain our own forks of other libraries, we need to do this right from the beginning so as not to develop bad habits that make it harder for us to maintain.

File: rq_dashboard/web.py [hunk 6]
```diff
@@ -81,10 +114,10 @@ def setup_rq_connection():
         current_app.redis_conn = sentinel.master_for(redis_master)
     else:
         current_app.redis_conn = Redis(
-            host=current_app.config.get('RQ_DASHBOARD_REDIS_HOST'),
-            port=current_app.config.get('RQ_DASHBOARD_REDIS_PORT'),
+            host=current_app.config.get('RQ_DASHBOARD_REDIS_HOST', 'localhost'),
```
Comment (lucasnad27) [inline rq_dashboard/web.py]: `app.config` should be responsible for setting default values, not the rq connection code

File: rq_dashboard/web.py [hunk 7]
```diff
@@ -340,5 +440,5 @@ def serialize_queue_names(worker):
 
 @blueprint.context_processor
 def inject_interval():
-    interval = current_app.config.get('RQ_DASHBOARD_POLL_INTERVAL')
+    interval = current_app.config.get('RQ_DASHBOARD_POLL_INTERVAL', 2500)
```
Comment (lucasnad27) [inline rq_dashboard/web.py]: Same here re: default values. The goal here is that I shouldn't have to grep for this all over the codebase to find multiple "default" values.

**Comment by ibrahim12** [comment]
> I noticed two commits from the upstream maintainer. We should _NOT_ spend our time reviewing the maintainer's code, unless we're looking at updating our fork. But that should be done outside of the context of our own changes and pull requests. Here is the recommended flow:
> 
> 1. git checkout master
> 2. git fetch --upstream
> 3. git merge upstream/master
> 
> If you have a feature branch that needs the most recent changes from the upstream branch, you should rebase* from origin/master (our version of master) after you've completed the steps from above.
> 
> * or merge if that's a foreign concept to you. Hint, hint...rebasing is a great skill in your tool belt to have!
> 
> Sorry I'm being a stickler on this, but if we're going to maintain our own forks of other libraries, we need to do this right from the beginning so as not to develop bad habits that make it harder for us to maintain.

hey @lucasnad27 I had forgot to update our master, I have updated our master, now no commits from upstream are in this PR to review.

File: rq_dashboard/web.py (hunk 7 above)
Comment (ibrahim12) [reply rq_dashboard/web.py]: It came from upstream.

File: rq_dashboard/web.py (hunk 6 above)
Comment (ibrahim12) [reply rq_dashboard/web.py]: It came from upstream.

Comment (lucasnad27) [reply rq_dashboard/web.py]: Ugh...at least you & I know better ;) We'll keep it as is in that case

**Review by lucasnad27 — APPROVED** [review]

**Pull request merged**